Fix unhandled promise rejection warning for validator

pjlee11 opened the pull request:
Resolves #674 

_Makes `expectMethodToBeCalledTimes` async and uses `await` to ensure the Promise rejection is handled. Also makes the test calls respect the `async/await`. Finally changes the expected values to be correct following the removal of the `fragmented` directory in the `/data`_

- [x] Tests added for new features
- [ ] Test engineer approval

## For testing:
### Should fail
1) `git checkout a157296fa6efca67a8a4f3bc0ace4e9a121b2be2`
2) `npm run test`
3) should see test correctly fail

### Should not fail
1) `git checkout fix-UnhandledPromiseRejectionWarning-for-validator`
2) `npm run test`
3) no test should fail

Commit:
fix validation warning

Commit:
Merge branch 'latest' of github.com:BBC-News/simorgh into fix-UnhandledPromiseRejectionWarning-for-validator

Commit:
make expectMethodToBeCalledTimes async to catch faliures

pjlee11 commented:
Blocked awaiting the refinement of #674

pjlee11 commented:
issue has been refined

ChrisBAshton commented on src/app/dataValidator/helpers/dataLoader/readdirSync.test.js:
```diff
@@ -6,14 +6,17 @@ const readScenario = require('./readScenario');
+  try {
     expect(spyMethod).toHaveBeenCalledTimes(number);
-  });
+  } catch (error) {
```
This looks as though you're expecting `expect(spyMethod).toHaveBeenCalledTimes(number);` to sometimes throw an exception...?

If you then call this again in the `catch` block, isn't it going to throw an exception again?

This is basically the same as:

```js
try { }
catch { }
finally {
  expect(spyMethod).toHaveBeenCalledTimes(number);
}
```

... which seems redundant! ^

ChrisBAshton replied:
Have now looked at this with Phil and yeah - it's complicated but kind of necessary! He's going to refactor slightly and add a clarifying comment.

pjlee11 replied:
Seeing this with fresh eyes the try catch isn't needed, good spot!

ChrisBAshton commented on src/app/dataValidator/helpers/dataLoader/readdirSync.test.js:
```diff
@@ -6,14 +6,17 @@ const readScenario = require('./readScenario');
   dataPath = defaultDataPath,
 ) => {
-  readdirSync(dataPath).then(() => {
+  await readdirSync(dataPath);
```
`readdirSync` is synchronous - why would it need `await`ing?

pjlee11 replied:
```
module.exports.readdirSync = dirName => {
  const fileNames = fs.readdirSync(dirName);

  return Promise.all(
    fileNames.map(fileName => readScenario.readScenario(fileName, dirName)),
  );
};
```

On review the method name is misleading, the function `readdirSync` actually does 2 things, synchronously reads the directory and then asynchronously validates each of the payloads returned from the synchronous directory read. 

Would renaming the method to `asyncValidateFilesInDirectory` clarify it's behaviour?

ChrisBAshton replied:
Ha, interesting! `readdirSync` -> `readdirAsync` would probably be fine.

Any reason why we've written our own implementation and not used the [native Node methods](https://nodejs.org/api/fs.html#fs_fs_readdirsync_path_options) or a helper like [fs-extra](https://www.npmjs.com/package/fs-extra)? The less we have to write & test ourselves, the better IMO - I can raise an issue to refactor these out in the future...?

pjlee11 replied:
Haven't looked into `fs-extra` before and IIRC we aren't using `fs.readdir` as it is easier to test the downstream `readScenario.readScenario` function by passing it an array rather than mocking a directory.

ChrisBAshton commented on src/app/dataValidator/helpers/dataLoader/readdirSync.test.js:
```diff
@@ -27,22 +30,22 @@ describe('readdirSync helper', () => {
     const readScenarioSpy = jest.spyOn(readScenario, 'readScenario');
 
-    expectMethodToBeCalledTimes(79, readScenarioSpy);
+    await expectMethodToBeCalledTimes(44, readScenarioSpy);
```
How was [this commit](https://github.com/BBC-News/simorgh/pull/904/commits/90da192c4114e42a6deb0e06134e2f4283b55672) passing before [this commit](https://github.com/BBC-News/simorgh/pull/904/commits/978dad7ef9cf380117d863536b35daf72f78f6cc)?

Looks as though I could change the expected value here and the PR would still have a green tick next to it. Can we make sure these tests are running?

pjlee11 replied:
That is the correct, the whole point of the issue is this test is failure on latest as the expected failures are incorrect. The test fails as per the details in the issue (`UnhandledPromiseRejectionWarning`).

pjlee11 replied:
The check the tests are now accurate, if check out the branch and `npm run test:unit` it'll pass with `44` and correctly error on anything other than `44` - which is not the case on latest as it silently fails

ChrisBAshton replied:
Awesome 👍in these cases I like to do a:

```
git commit -m "deliberately break the build to provide tests work"
...
git revert [the commit]
```

Can you make another commit that gives peace of mind? :-)

pjlee11 replied:
Good idea, have done this and added the commit to the description for testing purposes

ChrisBAshton reviewed (commented):
Made some comments.

Commit:
readdirSync -> asyncValidateFilesInDirectory

On review this method name is very misleading and therefore I'm changing it to be accurate

Commit:
remove unnecessary try catch block

Commit:
deliberately break the build to provide tests work

Commit:
Revert "deliberately break the build to provide tests work"

This reverts commit a157296fa6efca67a8a4f3bc0ace4e9a121b2be2.

ChrisBAshton commented on src/app/dataValidator/helpers/dataLoader/readScenario.js:
```diff
@@ -2,13 +2,13 @@ const fs = require('fs');
 const { validateData } = require('../validators/validateData');
 const { countScenarios } = require('../../utilities/countScenarios');
-const readdirSync = require('./readdirSync');
+const asyncValidateFilesInDirectory = require('./asyncValidateFilesInDirectory');
```
Can this be `const { asyncValidateFilesInDirectory }  = ` like the lines above it? Would tidy up line 11.

pjlee11 replied:
I'm not fully able to explain way but I tried this previously be best practice reasons and I get the following error:
```
TypeError: asyncValidateFilesInDirectory is not a function
    at ifDirectoryThenValidateNestedFiles (/Users/phillee/Code/BBC/simorgh/src/app/dataValidator/helpers/dataLoader/readScenario.js:13:5)
    at Object.readScenario (/Users/phillee/Code/BBC/simorgh/src/app/dataValidator/helpers/dataLoader/readScenario.js:26:5)
...
```

Because of the above I changed to using `asyncValidateFilesInDirectory.asyncValidateFilesInDirectory()`

ChrisBAshton replied:
Hmmm looks like we've faced it before: https://github.com/BBC-News/simorgh/blob/3a1754c3bb966cba093dd1229ab5fd11fdbc7615/src/app/dataValidator/helpers/dataLoader/readdirSync.js#L2

Fine, ignore for now!

ChrisBAshton commented on src/app/dataValidator/helpers/dataLoader/asyncValidateFilesInDirectory.test.js:
```diff
@@ -27,25 +32,27 @@ describe('readdirSync helper', () => {
   it('should return a promise', () => {
-    expect(readdirSync('./data') instanceof Promise).toEqual(true);
+    expect(asyncValidateFilesInDirectory('./data') instanceof Promise).toEqual(
+      true,
```
Argh, such a shame this now has to be on multiple lines! Maybe if we renamed to `readdirAsync` it would fit onto one line?

pjlee11 replied:
Looking at the method's responsibility I feel like it should clarify that it is actually running the validator code, but I do agree the naming is too long. Would `asyncValidateDir` be a sensible naming in your opinion?

ChrisBAshton reviewed (dismissed):
Approved, bar a few minor typos/fixes.

Commit:
Merge branch 'latest' of github.com:BBC-News/simorgh into fix-UnhandledPromiseRejectionWarning-for-validator

Commit:
rename asyncValidateFilesInDirectory to asyncValidateDir

Commit:
Merge branch 'latest' of github.com:BBC-News/simorgh into fix-UnhandledPromiseRejectionWarning-for-validator

Commit:
files also need rename

jtart reviewed (approved):
The tests should use their own schema and fixture files rather than files we use our application, this would mitigate the need to update the tests when the data/schema changes, but that's a fairly big piece of work so doesn't need to be considered here.

jamesbrumpton committed:
Merge branch 'latest' into fix-UnhandledPromiseRejectionWarning-for-validator

jamesbrumpton commented:
LGTM

jtart committed:
Merge branch 'latest' into fix-UnhandledPromiseRejectionWarning-for-validator

Commit:
Merge branch 'latest' into fix-UnhandledPromiseRejectionWarning-for-validator
